### PR TITLE
Support ppc64le online host upgrade

### PR DESCRIPTION
**Pull request by waynechen55**
* **New** subroutine ```validate_repo_arch``` in module ```lib/repo_tools.pm``` checks whether provided repository uri is valid for the architecture, for example:
  * **If** given ```repo_uri``` already has ```repo_arch```, for example: ```http://openqa.suse.de/assets/repo/fixed/SLE-15-SP4-Full-ppc64le-GM-Media1```,  check whether this ```repo_uri``` is valid by do http request. If this ```repo_uri```
is valid, return ```ok(1)```.
  * **If** given ```repo_uri``` does not have ```repo_arch```, for example: ```https://developer.download.nvidia.com/compute/cuda/repos/sles16/```, check whether it has architecture specific sub-folders by do http request to ```repo_uri/x86_64```. If ```repo_uri/x86_64``` is valid, check the corresponding ```repo_uri/repo_arch``` and ```repo_uri/noarch``` to do the same validation. Return ```ok(1)``` if the ```repo_uri``` does not have any architecture specific sub-folder or has valid ```repo_arch``` or ```noarch``` specific sub-folder.

  * **A** failure example which examined non-existent cuda repository on ppc64le:
ppc64le_cuda

* **Load** module ```tests/virt_autotest/restore_upgrade_system.pm``` to do additional restoration before doing host upgrade which introduces two subroutines:
  * **New** ```remove_installed_packages``` in ```lib/utils.pm``` to remove unnecessary or unsupported software packages, for example, sshpass on ppc64le (if it is installed before).
  * **Existing** ```subscribe_extensions_and_modules``` to deregister unnecessary or unsupported extensions, for example, PackageHub (if it is registered before).
  
* **Adapting** two modules ```lib/guest_installation_and_configuration_base.pm``` and ```lib/opensusebasetest.pm``` to facilitate image downloading and grub detecting on ppc64le.

* **Modify** yaml schedule ```schedule/virt_autotest/sle16_host_upgrade.yaml``` to support ppc64le online host upgrade, including:
  * **Conditional** section ```{{restore_before_upgrade}}``` to load module ```virt_autotest/restore_upgrade_system``` if ```ADDITIONAL_RESTORATION``` is ```1```.
  * **Corresponding** conditional section ```{{post_migration_validation}}``` to load ```virt_autotest/prepare_non_transactional_server``` to redo preparation if ```ADDITIONAL_RESTORATION``` is ```1```.
  * **Conditional** section ```{{install_host_ppc64le}}``` to install ppc64le host.

* **New** SLES16.0 ppc64le guest profile ```sles_16_traditional_64_kvm_hvm_ppc64le_agama_online_iso.xml```.

* **Please** also review https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/26486 which adds wider console selecting support for host upgrade.

* **Verification Runs:**
  * [ppc64le host upgrade from 16.0 to 16.1](https://openqa.suse.de/tests/24078597)
  * [sle-16.1-Online-aarch64-Build63.1-virt-host_upgrade-15sp7-to-16.1-ipmi](https://openqa.suse.de/tests/24118861)
  * [sle-16.1-Online-aarch64-Build63.1-virt-host_upgrade-16.0-to-16.1-ipmi](https://openqa.suse.de/tests/24078601)
  * [sle-16.1-Online-x86_64-Build63.1-virt-host_upgrade-15sp7-to-16.1-qemu](https://openqa.suse.de/tests/24078602)
  * [sle-16.1-Online-x86_64-Build63.1-virt-host_upgrade-16.0-to-16.1-qemu](https://openqa.suse.de/tests/24078603)
  * [sle-16.1-Online-x86_64-Build63.1-gi-guest_oraclelinux-on-host_developing-kvm](https://openqa.suse.de/tests/24078636)
  * [ sle-16.1-Online-x86_64-Build63.1-immutable-gi-online-guest_developing-on-host_developing-kvm](https://openqa.suse.de/tests/24078641)
  * [sle-16.1-Online-x86_64-Build63.1-uefi-gi-full-guest_developing-on-host_developing-kvm](https://openqa.suse.de/tests/24078642)
  * [sle-16.1-Online-x86_64-Build63.1-uefi-gi-guest_developing-on-host_sles15sp7-kvm](https://openqa.suse.de/tests/24078643)
  * [sle-16.1-Online-aarch64-Build63.1-sles_migration_15sp5_textmode_all_modules_and_extensions](https://openqa.suse.de/tests/24080612)
  * [sle-16.1-Online-aarch64-Build63.1-slmicro_migration](https://openqa.suse.de/tests/24080615)
  * [sle-16.1-Online-ppc64le-Build63.1-sles_migration_15sp5_ha](https://openqa.suse.de/tests/24080616)
  * [sle-16.1-Online-ppc64le-Build63.1-sles_migration_160_ha](https://openqa.suse.de/tests/24080620)
  * [sle-16.1-Online-ppc64le-Build63.1-sles_sap_migration_15sp6](https://openqa.suse.de/tests/24080621)
  * [sle-16.1-Online-s390x-Build63.1-sles_migration_15sp5_textmode](https://openqa.suse.de/tests/24080622)
  * [sle-16.1-Online-x86_64-Build63.1-sles_migration_15sp5_textmode](https://openqa.suse.de/tests/24080627)
  * [sle-16.1-Online-x86_64-Build63.1-slmicro_migration](https://openqa.suse.de/tests/24080637)
  * [sle-16.1-Online-x86_64-Build63.1-sles_migration_15sp7_textmode_rmt](https://openqa.suse.de/tests/24101943)
  * [sle-16.1-Online-aarch64-Build63.1-sles_migration_160_rmt](https://openqa.suse.de/tests/24101975)
  * [sle-16.1-Online-ppc64le-Build65.2-sles_migration_15sp6_textmode_all_modules_and_extensions](https://openqa.suse.de/tests/24118835)
  * [sle-16.1-Online-s390x-Build65.2-sles_migration_160_ha](https://openqa.suse.de/tests/24118836)
  * [sle-16.1-Online-x86_64-Build65.2-sles_migration_160_rmt](https://openqa.suse.de/tests/24118837)
  * [sle-16.1-Online-aarch64-Build65.2-sles_migration_15sp6_textmode_all_modules_and_extensions](https://openqa.suse.de/tests/24118838)

### CODE REVIEW

File: tests/console/validate_repos.pm [hunk 1]
```diff
@@ -9,13 +9,15 @@
...
+use repo_tools qw(validate_repo_properties parse_repo_arch);
 use scheduler 'get_test_suite_data';
+use utils 'select_backend_console';
+use Utils::Architectures 'is_ppc64le';
 
 sub run {
     my $test_data = get_test_suite_data();
 
-    select_console 'root-console';
+    select_backend_console(init => 0);
```
Comment (jknphy) [inline tests/console/validate_repos.pm]: Does this work for all backends? s390x zkvm or s390x zVM even. Here I pick some sample for your verifications: https://openqa.suse.de/tests/overview?flavor=Online&test=sles_default_standard_unattended&modules=validate_repos&distri=sle&version=16.1&build=63.1&groupid=612

File: tests/console/validate_repos.pm [hunk 2]
```diff
@@ -29,6 +31,11 @@ sub run {
     script_output 'zypper -n lr --uri';
 
     foreach my $repo (@{$test_data->{repos}}) {
+        if (is_ppc64le) {
+            my $parse_repo_arch = parse_repo_arch(repo_uri => $repo->{uri});
+            next if ($parse_repo_arch->{$repo->{uri}} == 2);
+            die if ($parse_repo_arch->{$repo->{uri}} == 3);
+        }
```
Comment (jknphy) [inline tests/console/validate_repos.pm]: I am not sure what you do here, like skip or die depending on some values. I would like to understand what is your plan here but at first glance make the maintenance of the module more complicated, perhaps not scheduling the module or scheduling a different squad-specific module here would help if it is not generic functionality, but need more context.

File: tests/yam/validate/validate_product.pm [hunk 3]
```diff
@@ -11,9 +11,10 @@ use testapi;
 use Config::Tiny;
 use Test::Assert ':all';
 use scheduler 'get_test_suite_data';
+use utils 'select_backend_console';
 
 sub run {
-    select_console 'root-console';
+    select_backend_console(init => 0);
```
Comment (jknphy) [inline tests/yam/validate/validate_product.pm]: I am wondering why the function needs to set parameter for qemu, which is the way that developers in bugs check first stuff, the quickest way to test something, also for qe regardless of the squad scope, I would say it is the default, remote workers it is kind of the special case, usually less test coverage, for those cases IMO the function should  needs those extra parameters, not for qemu. Additionally perldoc could be used to document each parameter of that function, it would be more clear.

Another concern about this function is why not set it at the level of the project instead of being importing it in each file?

Comment (jknphy) [reply tests/yam/validate/validate_product.pm]: Actually I think it would be a good idea to put this function in a different PR for wider audience as seems to have a wider scope.

File: tests/console/validate_repos.pm (hunk 2 above)
Comment (waynechen55) [reply tests/console/validate_repos.pm]: Will add more info later.

File: tests/yam/validate/validate_product.pm (hunk 3 above)
Comment (waynechen55) [reply tests/yam/validate/validate_product.pm]: Maybe a good idea, I will think and see what I can do.

File: tests/console/validate_repos.pm (hunk 2 above)
Comment (waynechen55) [reply tests/console/validate_repos.pm]: I can give you a quick introduction. What I want to do here is some repositories might not be valid for all architectures which need to be checked, for example, checking both the base url (repo_url)and arch specific url (repo_url/arch) before validating repositories. If directly provided repo_url/arch does not exist or repo_url/other_arch exits but repo_url/this_arch does not exist, skip validating corresponding repository. I want this to be generic, so basically is_ppc64le can be omitted.

File: tests/console/validate_repos.pm (hunk 1 above)
Comment (waynechen55) [reply tests/console/validate_repos.pm]: select_backend_console now in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/26486

File: tests/yam/validate/validate_product.pm (hunk 3 above)
Comment (waynechen55) [reply tests/yam/validate/validate_product.pm]: select_backend_console now in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/26486

File: tests/console/validate_repos.pm (hunk 2 above)
Comment (waynechen55) [reply tests/console/validate_repos.pm]: I updated the pull request and will continue to add pull request description and commit message. Please check back later and continue your review. Thanks.

**Comment by waynechen55** [comment]
Welcome your review @alice-suse @guoxuguang @Julie-CAO @RoyCai7 @nanzhg @tbaev @martinsmarcelo Please help have a look if you are interested and leave a message.

File: tests/console/validate_repos.pm [hunk 4]
```diff
@@ -29,6 +29,11 @@ sub run {
     script_output 'zypper -n lr --uri';
 
     foreach my $repo (@{$test_data->{repos}}) {
+        my $parse_repo_arch = '';
+        if ($repo->{uri}) {
+            $parse_repo_arch = parse_repo_arch(repo_uri => $repo->{uri});
```
Comment (jknphy) [inline tests/console/validate_repos.pm]: `sub parse_repo_arch` adds a tremendous complexity for what this module is not designed, I am wondering why is not under yam/ I guess because it was used in sles 15, but the origin of this module is just to do a simple check of the repos after installation.

Could you please think another way to implement this without touching this module?
Is it possible to use `skip_aliases` like above or use conditional schedule to not schedule this validation in power? I remember that this kind of test test-data-oriented test module only try to validate in a simple way what is passed by test data, not doing complex manipulation to guess values. A failure in that new library would be tricky to debug and also it is using http calls, it is quite a mechanism :)  Although it is true I don't understand exactly for which case you want to do that to suggest better, which reinforce my point about what would be the difficulty that a openQA reviewer would do in case of failure.

Comment (jknphy) [reply tests/console/validate_repos.pm]: Ahm now I see your example in the first comment in the PR:
- https://openqa.suse.de/assets/repo/fixed/SLE-15-SP4-Full-ppc64le-GM-Media1/
- https://developer.download.nvidia.com/compute/cuda/repos/sles16/x86_64/
why not be more specific with the architecture in the test data which admit interpolation with other setting like `%ARCH%` for instance? I don't see in your verification run any nvidia example, which is a very special/corner case.

File: lib/utils.pm [hunk 5]
```diff
@@ -3357,6 +3358,34 @@ sub install_extra_packages {
...
+different purposes, for example, restoring environment or unsupported scenario.
+User can specify pacakges to be removed via named argument packages which support
+multiple packages separated by comma. 
+=cut
+
+sub remove_installed_packages {
+    my %args = @_;
+    $args{packages} //= '';
+    if (!$args{packages}) {
+        record_info('No packags to be removed', 'Specify arguments packages');
```
Comment (RoyCai7) [inline lib/utils.pm]: s/packags/packages/

File: tests/virt_autotest/restore_upgrade_system.pm [hunk 6]
```diff
@@ -11,11 +11,16 @@
...
 use Mojo::Base 'opensusebasetest';
 use testapi;
-use virt_autotest::utils qw(select_backend_console reset_network_config);
+use virt_autotest::utils qw(select_backend_console reset_network_config subscribe_extensions_and_modules);
+use utils qw(remove_installed_packages);
 
 sub run {
     # Login as root
     select_backend_console(init => 0) if (!check_screen('text-logged-in-root'));
+    # Remove unecessary or unsupported packages
```
Comment (RoyCai7) [inline tests/virt_autotest/restore_upgrade_system.pm]: s/unecessary/unnecessary/

File: tests/virt_autotest/restore_upgrade_system.pm [hunk 7]
```diff
@@ -11,11 +11,16 @@
...
+    remove_installed_packages(packages => get_var('INSTALL_OTHER_PACKAGES', ''));
+    # Deregister unecessary or unsupported extensions
```
Comment (RoyCai7) [inline tests/virt_autotest/restore_upgrade_system.pm]: s/unecessary/unnecessary/

File: data/virt_autotest/guest_params_xml_files/sles_16_traditional_64_kvm_hvm_ppc64le_agama_online_iso.xml [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  <guest_boot_settings/>
+  <guest_xpath/>
+  <guest_installation_automation_method>autoagama</guest_installation_automation_method>
+  <guest_installation_automation_platform/>
+  <guest_installation_automation_file>sles_16_plus_kvm_hvm_guest_non_encrypted_storage_ppc64le.jsonnet</guest_installation_automation_file>
+  <guest_installation_method>directkernel</guest_installation_method>
+  <guest_installation_extra_args/>
+  <guest_installation_wait/>
+  <guest_installation_method_others/>
+  <guest_installation_media>http://openqa.suse.de/assets/iso/fixed/SLES-16.0-Online-ppc64le-GM.install.iso</guest_installation_media>
```
Comment (alice-suse) [inline data/virt_autotest/guest_params_xml_files/sles_16_traditional_64_kvm_hvm_ppc64le_agama_online_iso.xml:29]: We are moving majority of existing tests from online to full for sle16+. For new added tests, it will be good to directly use full to save time.

File: lib/guest_installation_and_configuration_base.pm [hunk 9]
```diff
@@ -1511,8 +1511,8 @@ sub config_guest_installation_method {
         my $_guest_arch = ($self->{guest_arch} ? $self->{guest_arch} : get_required_var('ARCH'));
         if (script_output("curl --silent -I $_guest_installation_fine_grained_media | grep -E \"^HTTP\" | awk -F \" \" \'{print \$2}\'") == "200") {
             if ($self->{guest_installation_method} eq 'directkernel') {
-                assert_script_run("curl -s -o $self->{guest_image_folder}/linux $_guest_installation_fine_grained_media/boot/$_guest_arch/loader/linux");
-                assert_script_run("curl -s -o $self->{guest_image_folder}/initrd $_guest_installation_fine_grained_media/boot/$_guest_arch/loader/initrd");
+                assert_script_run("curl -s -o $self->{guest_image_folder}/linux $_guest_installation_fine_grained_media/boot/$_guest_arch/" . (is_ppc64le ? "linux" : "loader/linux"));
```
Comment (alice-suse) [inline lib/guest_installation_and_configuration_base.pm]: Just curious, why such code change not needed in current 16.1 GI test? Seems it is critical need.

File: lib/repo_tools.pm [hunk 10]
```diff
@@ -562,6 +564,69 @@ sub validate_repo_properties {
...
+check corresponding repo_uri/repo_arch to do the same validation. In this case,
+repo_arch_uri=repo_uri/repo_arch and repo_base_uri=repo_uri.
+The returned values are 0 or 1 which represents in/valid repository uri for the
+architecture involved.
+C<$args> should have following keys defined:
+- C<repo_uri>: The URIs of repositories which support multiple URIs separated by
+               comma or space.
+=cut
+
+sub parse_repo_arch {
```
Comment (alice-suse) [inline lib/repo_tools.pm]: Could the purpose of the function be reflected in the func name more -- Validates whether repositories are valid for specific architecture?

File: schedule/virt_autotest/sle16_host_upgrade.yaml [hunk 11]
```diff
@@ -70,13 +89,17 @@ conditional_schedule:
...
         - yam/migration/zypper_migration
   post_migration_boot:
     AUTOYAST_PREPARE_PROFILE:
       "1":
         - installation/grub_test
         - installation/first_boot
+  post_migration_validation:
+    ADDITIONAL_RESTORATION:
+      "1":
+        - virt_autotest/prepare_non_transactional_server
```
Comment (alice-suse) [inline schedule/virt_autotest/sle16_host_upgrade.yaml:101]: This file rename will directly impact host upgrade test on existing arches. Let's hold on merge until the job group MR is ready too.

File: tests/virt_autotest/restore_upgrade_system.pm [hunk 12]
```diff
@@ -11,11 +11,16 @@
 
 use Mojo::Base 'opensusebasetest';
 use testapi;
-use virt_autotest::utils qw(select_backend_console reset_network_config);
+use virt_autotest::utils qw(select_backend_console reset_network_config subscribe_extensions_and_modules);
+use utils qw(remove_installed_packages);
 
 sub run {
     # Login as root
     select_backend_console(init => 0) if (!check_screen('text-logged-in-root'));
```
Comment (alice-suse) [inline tests/virt_autotest/restore_upgrade_system.pm:19]: This init => 0 not necessary now?

**Review by alice-suse — COMMENTED** [review]
Thanks for making it work! I read it through and left some minor comments.

File: data/virt_autotest/guest_params_xml_files/sles_16_traditional_64_kvm_hvm_ppc64le_agama_online_iso.xml (hunk 8 above)
Comment (alice-suse) [reply data/virt_autotest/guest_params_xml_files/sles_16_traditional_64_kvm_hvm_ppc64le_agama_online_iso.xml:29]: Ignore this, QU will not need upgrade test, so okay to be either case.

File: tests/console/validate_repos.pm (hunk 4 above)
Comment (waynechen55) [reply tests/console/validate_repos.pm]: 1. Why do you think ```parse_repo_arch``` is tremendous complex ? Its logic is pretty simple, just querying the uri and nothing else. It just differentiates several different cases with regard to uri, including: 
```uri without architecture in uri but have involved architecture sub-folder```
```uri without architecture in uri but have x86 sub-folder and does not involved architecture sub-folder```
```uri without architecture in uri and does not have architecture sub-folder```
```uri has involved architecture in it```.

2. I tried verifying with ```test_data/yam/agama_sle.yaml``` and ```test_data/yam/agama_sle_extended_unattended_s390x_ppc64le.yaml``` both of which failed actually.

3. What is your suggested fix in details ? 

4. If I also add ```is_ppc64le``` and other conditions to further limit the change, do you think it is acceptable ?

@jknphy

File: lib/guest_installation_and_configuration_base.pm (hunk 9 above)
Comment (waynechen55) [reply lib/guest_installation_and_configuration_base.pm]: ppc64le 16.0 and 16.1 media has different folder hierarchy. I need to differentiate it further.

File: tests/virt_autotest/restore_upgrade_system.pm (hunk 12 above)
Comment (waynechen55) [reply tests/virt_autotest/restore_upgrade_system.pm:19]: init defaults to 0 in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/26486 now which can simplify function calling.

File: lib/utils.pm (hunk 5 above)
Comment (waynechen55) [reply lib/utils.pm]: Done

File: tests/virt_autotest/restore_upgrade_system.pm (hunk 6 above)
Comment (waynechen55) [reply tests/virt_autotest/restore_upgrade_system.pm]: Done

File: tests/virt_autotest/restore_upgrade_system.pm (hunk 7 above)
Comment (waynechen55) [reply tests/virt_autotest/restore_upgrade_system.pm]: Done

File: tests/console/validate_repos.pm (hunk 4 above)
Comment (waynechen55) [reply tests/console/validate_repos.pm]: Maybe introducing a variable in if condition ```if get_var('VALIDATE_REPO_ARCH')``` is far better. HDYT ? @jknphy

Comment (jknphy) [reply tests/console/validate_repos.pm]: let's solve a simple question first if you might please:
Is it possible to setup in test data that string for those 4 uri variants you mentioned above using %ARCH% as interpolation or duplicating the test data files? Does the current tooling/code allow you to do that?

Comment (jknphy) [reply tests/console/validate_repos.pm]: My initial concerns about complexity (I didn't check it in detail the function yet) are that you are doing network calls, and this kind of validation tests should just use a library in Perl or some assert library to check against FIXED test data (that is the reason of my question above). That is the original design and idea as far as I remember. This implementation deviates from that.

Comment (jknphy) [reply tests/console/validate_repos.pm]: what comes to my mind if it would not be possible to setup this with current tooling is to 'invent' new field in test data that you could use to decide further what to do, but you always know on advance in test data what you are looking for instead of driving the test depending on the result of the url check, so we would not break the test-data-driven approach this validation test module is using, avoiding dynamic behaviors.

Comment (waynechen55) [reply tests/console/validate_repos.pm]: 1. I think ```parse_repo_arch``` (will be renamed to ```validate_repo_arch``` maybe) is pretty useful from my POV. I tend to have this function in ```lib/repo_tools.pm```. It is more like a generic subroutine which can be called anywhere regardless of test scenarios (although might not be used at the moment).

2. I think introducing a ```ARCH``` in ```test_data.yaml``` might be possible. So ```validate_repo_properties``` will also check ```ARCH``` before continuing. Do you think it is acceptable ? If yes, I will continue to have a deeper look. @jknphy

Comment (jknphy) [reply tests/console/validate_repos.pm]: could you please provide me a simple example of one case that the current code cannot do?
which verification can I check and what is the output of the existing repos that need to be parsed?
(2) would be the the best way but without updating this file or validate_repo_properties, just test data.

Comment (waynechen55) [reply tests/console/validate_repos.pm]: The example is with either ```test_data/yam/agama_sle_extended_unattended_s390x_ppc64le.yaml``` or ```test_data/yam/agama_sle.yaml```, repository validating failed at following:
image

I do not have failed job run link anymore or you can visit [this link](http://10.144.179.17/tests/1873#step/validate_repos/13), but I can not guarantee you have access.

Comment (jknphy) [reply tests/console/validate_repos.pm]: when I set in my browser url https://static.scc.suse.com/repos/cuda/sles16/x86_64 got redirected to https://developer.download.nvidia.com/compute/cuda/repos/sles16/x86_64/ 
I cannot access that link as you mentioned.

Why work for us and not for you? do you have different test data?
https://openqa.suse.de/tests/23891035#step/validate_repos/5

Comment (waynechen55) [reply tests/console/validate_repos.pm]: Your passed example is aarch64 :)

Comment (jknphy) [reply tests/console/validate_repos.pm]: exactly, we pass a different test data for ppc64le because there is not power for NVIDIA, and that is exactly how a test-data-driven module should behaves: https://openqa.suse.de/tests/23900273/settings/test_data/yam/agama_sle_s390x_ppc64le.yaml
Test data is the source of true, the method to validate should not have any other logic than validate it, match expected vs actual results.

Comment (waynechen55) [reply tests/console/validate_repos.pm]: I will try this one ```agama_sle_s390x_ppc64le.yaml```.

File: lib/guest_installation_and_configuration_base.pm (hunk 9 above)
Comment (alice-suse) [reply lib/guest_installation_and_configuration_base.pm]: I see. But why is here no if else to differentiate the two products and only ppc check?

File: tests/virt_autotest/restore_upgrade_system.pm (hunk 12 above)
Comment (alice-suse) [reply tests/virt_autotest/restore_upgrade_system.pm:19]: yes, so could remove init -> 0 here. That's what I meant.

File: lib/guest_installation_and_configuration_base.pm (hunk 9 above)
Comment (waynechen55) [reply lib/guest_installation_and_configuration_base.pm]: Yes. I will do it.

File: tests/virt_autotest/restore_upgrade_system.pm (hunk 12 above)
Comment (waynechen55) [reply tests/virt_autotest/restore_upgrade_system.pm:19]: It is removed in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/26486

File: tests/console/validate_repos.pm (hunk 4 above)
Comment (waynechen55) [reply tests/console/validate_repos.pm]: Test run with ```agama_sle_s390x_ppc64le.ymal``` passed ```validate_repos.pm```. @jknphy

Comment (jknphy) [reply tests/console/validate_repos.pm]: Glad that was as simple as using test data, thanks for taking a look.
I guess you plan to drop `validate_repo_arch` as there is not used of this function now and dead code is code hard to maintain unless you find some other space/file more clear for experimentation, but I leave that up to you.

File: lib/guest_installation_and_configuration_base.pm (hunk 9 above)
Comment (waynechen55) [reply lib/guest_installation_and_configuration_base.pm]: Done. [non ppc64le guest](https://openqa.suse.de/tests/24190047), [15sp7 guest](https://openqa.suse.de/tests/24190048), [16.0 guest](https://openqa.suse.de/tests/24190049) and [16.1 guest](https://openqa.suse.de/tests/24190050) all passed.

File: lib/repo_tools.pm (hunk 10 above)
Comment (waynechen55) [reply lib/repo_tools.pm]: Done

**Comment by waynechen55** [comment]
All additional verification runs passed:

[ppc64le on host upgrade](https://openqa.suse.de/tests/24273648)
[ sle-16.1-Online-x86_64-Build66.1-immutable-gi-online-guest_developing-on-host_developing-kvm](https://openqa.suse.de/tests/24190047)
[sle-16.1-Online-x86_64-Build66.1-sev-snp-guest_sles15sp7-on-host_developing-kvm](https://openqa.suse.de/tests/24190048)
[sle-16.1-Online-x86_64-Build66.1-snapshot-gi-guest_sles16_0-online-on-host_developing-kvm](https://openqa.suse.de/tests/24190049)
[le-16.1-Online-x86_64-Build66.1-uefi-gi-guest_developing-on-host_sles16_0-kvm](https://openqa.suse.de/tests/24190050)
[ sle-16.1-Online-aarch64-Build66.1-uefi-gi-guest_postsles16-on-host_developing-kvm](https://openqa.suse.de/tests/24190051)
[sle-16.1-Online-aarch64-Build66.1-uefi-gi-guest_presles16-on-host_developing-kvm](https://openqa.suse.de/tests/24190053)
[sle-16.1-Online-aarch64-Build66.1-virt-host_upgrade-15sp7-to-16.1-ipmi](https://openqa.suse.de/tests/24204501)
[sle-16.1-Online-x86_64-Build67.2-virt-host_upgrade-15sp7-to-16.1-qemu](https://openqa.suse.de/tests/24249389)
[sle-16.1-Online-aarch64-Build67.2-virt-host_upgrade-16.0-to-16.1-ipmi](https://openqa.suse.de/tests/24249394)
[sle-16.1-Online-x86_64-Build67.2-virt-host_upgrade-16.0-to-16.1-qemu](https://openqa.suse.de/tests/24249396)
[sle-16.1-Transactional-Base-aarch64-Build13.28](https://openqa.suse.de/tests/24204505)
[sle-16.1-Transactional-Base-SelfInstall-x86_64-Build13.28](https://openqa.suse.de/tests/24204506)
[sle-16.1-Transactional-Base-ppc-4096-ppc64le-Build13.28](https://openqa.suse.de/tests/24204507)
[sle-16.1-Transactional-Base-qcow-s390x-Build13.28](https://openqa.suse.de/tests/24204508)
[sle-16.1-Online-ppc64le-Build66.1-create_hdd_external_testkit](https://openqa.suse.de/tests/24204526)
[sle-16.1-Online-ppc64le-Build66.1-create_hdd_textmode](https://openqa.suse.de/tests/24204527)
[sle-16.1-Online-ppc64le-Build66.1-stack_tests_python](https://openqa.suse.de/tests/24204528)
[sle-16.1-Online-Immutable-ppc64le-Build66.1-create_hdd_textmode](https://openqa.suse.de/tests/24204529)
[sle-16.1-Online-Immutable-ppc64le-Build66.1-toolchain_zypper](https://openqa.suse.de/tests/24204530)
[sle-16.1-Online-ppc64le-Build66.1-io_uring](https://openqa.suse.de/tests/24204532)
[sle-16.1-Online-ppc64le-Build66.1-kernel-live-patching](https://openqa.suse.de/tests/24204533)
[sle-16.1-Online-ppc64le-Build66.1-sles_migration_15sp7_textmode](https://openqa.suse.de/tests/24204510)
[ sle-16.1-Online-ppc64le-Build66.1-sles_migration_15sp7_textmode](https://openqa.suse.de/tests/24244759)
[sle-16.1-Online-ppc64le-Build66.1-sles_migration_15sp5_textmode_all_modules_and_extensions](https://openqa.suse.de/tests/24204511)
[sle-16.1-Online-ppc64le-Build66.1-sles_sap_migration_15sp5](https://openqa.suse.de/tests/24204512)
[sle-16.1-Online-aarch64-Build66.1-sles_migration_15sp7_textmode_rmt](https://openqa.suse.de/tests/24204521)
[sle-16.1-Online-x86_64-Build66.1-sles_migration_15sp7_textmode](https://openqa.suse.de/tests/24204523)
[sle-16.1-Minimal-VM-for-MS-HyperV-x86_64-Build22.6-minimalvm-containers-docker](https://openqa.suse.de/tests/24246000)
[sle-16.0-VIRT-Kernel-Updates-Staging-x86_64-Build:git:7646:kernel-source:kernel-rt-mau-extratests-virt-hyperv](https://openqa.suse.de/tests/24246001)
[sle-15-SP7-JeOS-for-MS-HyperV-QR-x86_64-Build8.7-jeos-containers-fips-podman](https://openqa.suse.de/tests/24246002)
[sle-16.1-Full-s390x-Build67.2-gi-full-guest_developing-on-host_developing-kvm](https://openqa.suse.de/tests/24246003)
[sle-16.1-Base-SelfInstall-pxe-x86_64-Build13.31-gi-guest_developing-on-host_developing-kvm](https://openqa.suse.de/tests/24248533)
[sle-16.1-Base-SelfInstall-pxe-x86_64-Build13.31-gi-guest_slem6dot2-on-host_developing-kvm](https://openqa.suse.de/tests/24248535)
[sle-16.1-Base-SelfInstall-pxe-x86_64-Build13.31-gi-guest_sles15sp7-on-host_developing-kvm](https://openqa.suse.de/tests/24248543)
[sle-16.1-Transactional-Base-encrypted-x86_64-Build13.31-sles16_transactional_virtualization](https://openqa.suse.de/tests/24249510)
[sle-16.1-Transactional-Base-SelfInstall-aarch64-Build13.31-sles16_transactional_installation_virt](https://openqa.suse.de/tests/24249508)
[ sle-16.1-Online-aarch64-Build67.2-sles_migration_15sp5_textmode](https://openqa.suse.de/tests/24248590)
[sle-16.1-Online-ppc64le-Build67.2-sles_migration_15sp7_textmode](https://openqa.suse.de/tests/24248598)
[sle-16.1-Online-ppc64le-Build67.2-sles_migration_160_ha](https://openqa.suse.de/tests/24248601)
[sle-16.1-Online-s390x-Build67.2-sles_migration_15sp5_textmode](https://openqa.suse.de/tests/24248602)
[sle-16.1-Online-s390x-Build67.2-sles_migration_15sp7_textmode_scc](https://openqa.suse.de/tests/24248603)
[sle-16.1-Online-x86_64-Build67.2-sles_migration_15sp6_ha](https://openqa.suse.de/tests/24248608)

**Comment by waynechen55** [comment]
Ready for merge @alice-suse @guoxuguang @Julie-CAO @RoyCai7 @nanzhg @tbaev @martinsmarcelo Please help have a look if you are interested and leave a message.

**Review by alice-suse — APPROVED** [review]

**Review by Julie-CAO — APPROVED** [review]
